unison-src/transcripts/kind-inference.output.md

```diff
@@ -0,0 +1,289 @@
+
+## A type param cannot have conflicting kind constraints within a single decl
+
+conflicting constraints on the kind of `a` in a product
+```unison
+unique type T a = T a (a Nat)
+```
+
+```ucm
+
+  Kind mismatch arising from
+        1 | unique type T a = T a (a Nat)
+    
+    a doesn't expect an argument; however, it is applied to Nat.
+
+```
+conflicting constraints on the kind of `a` in a sum
+```unison
+unique type T a 
+  = Star a 
+  | StarStar (a Nat)
+```
+
+```ucm
+
+  Kind mismatch arising from
+        3 |   | StarStar (a Nat)
+    
+    a doesn't expect an argument; however, it is applied to Nat.
+
+```
+## Kinds are inferred by decl component
+
+Successfully infer `a` in `Ping a` to be of kind `* -> *` by
+inspecting its component-mate `Pong`.
+```unison
+unique type Ping a = Ping Pong
+unique type Pong = Pong (Ping Optional)
+```
+
+```ucm
+
+  I found and typechecked these definitions in scratch.u. If you
+  do an `add` or `update`, here's how your codebase would
+  change:
+  
+    ⍟ These new definitions are ok to `add`:
+    
+      unique type Ping a
+      unique type Pong
+
+```
+Catch the conflict on the kind of `a` in `Ping a`. `Ping` restricts
+`a` to `*`, whereas `Pong` restricts `a` to `* -> *`.
+```unison
+unique type Ping a = Ping a Pong
+unique type Pong = Pong (Ping Optional)
+```
+
+```ucm
+
+  Kind mismatch arising from
+        1 | unique type Ping a = Ping a Pong
+    
+    The arrow type (->) expects arguments of kind *; however, it
+    is applied to a which has kind: * -> *.
```

Comment:
In a future ticket we should improve this message which mentions arrow though we don't see any.